### PR TITLE
Eagerly read IAsyncEnumerable{object} instances before formatting

### LINKED ISSUE #4833: Add IAsyncEnumerable support to MVC
**Edit: @rynowak hijacking top post for great justice**

This would include adding support for returning `IAsyncEnumerable<>` from an action method and letting MVC do the buffering for you because giving the data to the formatter.

This would require integrating with JsonResult and ObjectResult

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes https://github.com/aspnet/AspNetCore/issues/4833

### CODE REVIEW

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 1]
```diff
@@ -18,16 +23,43 @@ namespace Microsoft.AspNetCore.Mvc.Infrastructure
     /// </summary>
     public class ObjectResultExecutor : IActionResultExecutor<ObjectResult>
     {
+        private delegate Task<object> ReadAsyncEnumerableDelegate(object value);
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: Do we need this delegate type? Why not just `Func<object, Task<object>>`

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 2]
```diff
@@ -138,5 +190,50 @@ private static void InferContentTypes(ActionContext context, ObjectResult result
...
         }
+
+        private Task<object> EnumerateAsyncEnumerable(IAsyncEnumerable<object> value)
+        {
+            var type = value.GetType();
+            if (!_asyncEnumerableConverters.TryGetValue(type, out var result))
+            {
+                var enumerableType = ClosedGenericMatcher.ExtractGenericInterface(type, typeof(IAsyncEnumerable<>));
+                result = null;
+                if (enumerableType != null)
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: Should this ever fail? This seems like you should just let it blow up right?

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs (hunk 1 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: It was ugly to repeat it. But I can do a `using`

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 3]
```diff
@@ -138,5 +190,50 @@ private static void InferContentTypes(ActionContext context, ObjectResult result
                 result.ContentTypes.Add("application/problem+xml");
             }
         }
+
+        private Task<object> EnumerateAsyncEnumerable(IAsyncEnumerable<object> value)
+        {
+            var type = value.GetType();
+            if (!_asyncEnumerableConverters.TryGetValue(type, out var result))
+            {
+                var enumerableType = ClosedGenericMatcher.ExtractGenericInterface(type, typeof(IAsyncEnumerable<>));
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: What happens if you implement multiple `IAsyncEnumerable<>` interfaces? implementation-defined? Crash with a bad error?

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs (hunk 2 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: Yup. Left over from before. I'll Debug.Assert it.

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 4]
```diff
@@ -138,5 +190,50 @@ private static void InferContentTypes(ActionContext context, ObjectResult result
...
+                result = null;
+                if (enumerableType != null)
+                {
+                    var enumeratedObjectType = enumerableType.GetGenericArguments()[0];
+
+                    var converter = (ReadAsyncEnumerableDelegate)Converter
+                        .MakeGenericMethod(enumeratedObjectType)
+                        .CreateDelegate(typeof(ReadAsyncEnumerableDelegate), this);
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: cheeky.

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 5]
```diff
@@ -138,5 +190,50 @@ private static void InferContentTypes(ActionContext context, ObjectResult result
...
+        private Task<object> EnumerateAsyncEnumerable(IAsyncEnumerable<object> value)
+        {
+            var type = value.GetType();
+            if (!_asyncEnumerableConverters.TryGetValue(type, out var result))
+            {
+                var enumerableType = ClosedGenericMatcher.ExtractGenericInterface(type, typeof(IAsyncEnumerable<>));
+                result = null;
+                if (enumerableType != null)
+                {
+                    var enumeratedObjectType = enumerableType.GetGenericArguments()[0];
```
Comment (pranavkm) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: Btw, we need to do this because XML serializer needs to be told the type of the list. Giving it a vanilla `List<object>` causes them to not be serialized correctly.

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: I should probably add a note

File: src/Mvc/Mvc.Core/src/Resources.resx [hunk 6]
```diff
@@ -503,5 +503,8 @@
   </data>
   <data name="Property_MustBeInstanceOfType" xml:space="preserve">
     <value>Property '{0}.{1}' must be an instance of type '{2}'.</value>
-  </data>  
-</root>
+  </data>
+  <data name="ObjectResultExecutor_MaxEnumerationExceeded" xml:space="preserve">
+    <value>{0} execeeded the maximum configured enumeration limit '{1}' when enumerating async enumerable type '{2}'.</value>
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Resources.resx]: Should we add more details here?

This doesn't address how a user would solve a problem if they have one.

I think it would be nice to say something like:

```
The '{formatter}' reached the configured maximum size of the buffer when an enumerating a value of type `FooType'. This limit is in place to prevent infinite streams of `IAsyncEnumerable<>` from continuing indefinitely. If this is not a programming mistake, consider ways to reduce the collection or size, or consider manually converting 'FooType' into a list rather than increasing the limit. 
```

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 7]
```diff
@@ -2,14 +2,19 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Concurrent;
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: JsonResult? You might want to convert all of this new logic to a shared file.

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs (hunk 1 above)
Comment (rynowak) [reply src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: There's no difference between this and a using. My point is that I find `Func<>` a lot more readable than custom delegate types. If you don't agree then feel free to ignore.

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs (hunk 7 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: yup yup. Totally forgot about Json

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs (hunk 1 above)
Comment (mburbea) [reply src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: I'd argue that the using doesn't create an extra type, so it's slightly better.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    /// <summary>
+    /// Type that reads an <see cref="IAsyncEnumerable{T}"/> instance into a
+    /// generic collection instance.
+    /// </summary>
+    /// <remarks>
+    /// This type is used to create a strongly typed synchronous <see cref="ICollection{T}"/> instance from
+    /// an <see cref="IAsyncEnumerable{T}"/>. An accurate <see cref="ICollection{T}"/> is required for XML formatters to 
+    /// correctly serialize.
+    /// </remarks>
+    public sealed class AsyncEnumerableReader
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs]: Does this really need to be public? I don't have a strong objection, but it seems like infrastructure, rather than something we expect users to need.

File: src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            var type = value.GetType();
+            if (!_asyncEnumerableConverters.TryGetValue(type, out var result))
+            {
+                var enumerableType = ClosedGenericMatcher.ExtractGenericInterface(type, typeof(IAsyncEnumerable<>));
+                Debug.Assert(enumerableType != null);
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs]: This is a public API, so this should probably throw instead of assert. Someone could pass in any type here.

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 10]
```diff
@@ -138,5 +177,7 @@ private static void InferContentTypes(ActionContext context, ObjectResult result
                 result.ContentTypes.Add("application/problem+xml");
             }
         }
+
+
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: @JamesNK

File: src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs [hunk 11]
```diff
@@ -87,16 +107,35 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
...
+                return ExecuteAsyncEnumerable(context, result, asyncEnumerable);
+            }
+
+            return ExecuteAsyncCore(context, result, objectType, value);
+        }
+
+        private async Task ExecuteAsyncEnumerable(ActionContext context, ObjectResult result, IAsyncEnumerable<object> asyncEnumerable)
+        {
+            var enumerated = await _asyncEnumerableReader.ReadAsync(asyncEnumerable);
+            await ExecuteAsyncCore(context, result, enumerated.GetType(), enumerated);
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/ObjectResultExecutor.cs]: Should we be logging here?

File: src/Mvc/Mvc.Core/src/Infrastructure/SystemTextJsonResultExecutor.cs [hunk 12]
```diff
@@ -123,11 +134,22 @@ private static class Log
                 new EventId(1, "JsonResultExecuting"),
                 "Executing JsonResult, writing value of type '{Type}'.");
 
+            private static readonly Action<ILogger, string, Exception> _eagerlyReadingAsyncEnumerable = LoggerMessage.Define<string>(
+               LogLevel.Debug,
+               new EventId(2, "EagerReadAsyncEnumerable"),
+               "Eagerly reading IAsyncEnumerable instance of type '{Type}'.");
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/SystemTextJsonResultExecutor.cs]: Can we say something like `Eagerly buffering` instead of reading? I think it's a little confusing to say that we're reading when we're about to write. I realize that it's technically correct, but still a little misleading.

File: src/Mvc/Mvc.Core/src/Infrastructure/SystemTextJsonResultExecutor.cs [hunk 13]
```diff
@@ -123,11 +134,22 @@ private static class Log
...
+
             public static void JsonResultExecuting(ILogger logger, object value)
             {
                 var type = value == null ? "null" : value.GetType().FullName;
                 _jsonResultExecuting(logger, type, null);
             }
+
+            public static void EagerlyReadingAsyncEnumerable(ILogger logger, object value)
+            {
+                var type = value == null ? "null" : value.GetType().FullName;
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Infrastructure/SystemTextJsonResultExecutor.cs]: The value can't really be null here right? This would fail the `is`-cast.

File: src/Mvc/Mvc.Core/src/Resources.resx [hunk 14]
```diff
@@ -503,5 +503,8 @@
   </data>
   <data name="Property_MustBeInstanceOfType" xml:space="preserve">
     <value>Property '{0}.{1}' must be an instance of type '{2}'.</value>
-  </data>  
-</root>
+  </data>
+  <data name="ObjectResultExecutor_MaxEnumerationExceeded" xml:space="preserve">
+    <value>'{0}' reached the configured maximum size of the buffer when an enumerating a value of type `{1}'. This limit is in place to prevent infinite streams of `IAsyncEnumerable&lt;&gt;` from continuing indefinitely. If this is not a programming mistake, consider ways to reduce the collection or size, or consider manually converting '{1}' into a list rather than increasing the limit.</value>
```
Comment (rynowak) [inline src/Mvc/Mvc.Core/src/Resources.resx]: `collection or size` => `collection size`

File: src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonResultExecutor.cs [hunk 15]
```diff
@@ -131,7 +134,13 @@ await using (var writer = _writerFactory.CreateWriter(responseStream, resolvedCo
                     jsonWriter.AutoCompleteOnClose = false;
 
                     var jsonSerializer = JsonSerializer.Create(jsonSerializerSettings);
-                    jsonSerializer.Serialize(jsonWriter, result.Value);
+                    var value = result.Value;
+                    if (result.Value is IAsyncEnumerable<object> asyncEnumerable)
+                    {
+                        value = await _asyncEnumerableReader.ReadAsync(asyncEnumerable);
```
Comment (rynowak) [inline src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonResultExecutor.cs]: Should we log here?

**Review by rynowak — APPROVED** [review]

File: src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs (hunk 8 above)
Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs]: I could defer reading this from DI or have duplicated caching if your app uses a mix of `ObjectResult` and `JsonResult` returning `IAsyncEnumerable`. I guess the latter is rare and easily fixable if it turns out to be a problem in the future.

Comment (rynowak) [reply src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs]: I don't understand your answer. What's the reason why this type is public?

Comment (pranavkm) [reply src/Mvc/Mvc.Core/src/Infrastructure/AsyncEnumerableReader.cs]: `ObjectResultExecutor` is public. If we have to share the same singleton instance between the JsonResultExecutor(s) and ObjectResultExecutor, we either have to make the type public for ctor DI or read it in the body of `ExecuteAsync`.
